display_list/skia/dl_sk_canvas.cc

```diff
@@ -49,6 +45,20 @@ static SkPaint ToSk(const DlPaint& paint, bool force_stroke = false) {
     color_filter = invert_filter;
   }
   sk_paint.setColorFilter(color_filter);
+
+  // On the Impeller backend, we will only support dithering of *gradients*,
+  // and it will be enabled by default (without the option to disable it). Until
+  // Skia support is completely removed, we only want to respect the dither flag
+  // for gradients (otherwise it will also apply to, for example, images, which
+  // is not supported in Impeller).
+  //
+  // See https://github.com/flutter/flutter/issues/112498.
+  auto color_source = paint.getColorSourcePtr();
+  sk_paint.setShader(ToSk(color_source));
```

Comment:
Could move this inside the if. What was wrong with my original pseudo-code?

Also, sk_paint.setDither should be based on whether or not the color source is a gradient and not on paint.isDither()...

```
auto color_source = paint.getColorSourcePtr();
if (color_source) {
  sk_paint.setDither(color_source->isGradient());
  sk_paint.setShader(ToSk(color_source));
}
```

Reply:
Done (inside the `if`).

For your second question, no, that's not the behavior we want. It's a bit convoluted, but we want to set dithering IFF:
1. the source is a gradient
1. dithering is requested

Dithering will soon be enabled by _default_ (it's currently opt-in), but we will still allow users to opt-out for a period fo time.

Reply:
Ah, future tense. I think I missed that...

So, first pass, only set dither inside color_source handling.
Second pass, if making it the default sticks, eliminate the option and the method...